### DISCUSSION TITLE
getting_started/first_3d_game/going_further

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#103] [User-contributed notes]
# getting_started/first_3d_game/going_further

You can pat yourself on the back for having completed your first 3D game with Godot. In this series, we went over a wide range of techniques and editor features. Hopefully, you've witnessed how int...

https://docs.godotengine.org/en/stable/getting_started/first_3d_game/going_further.html

**Comment by VinceCefalu** [comment]
Thanks for a tutorial guys! It was easy to follow and introduced some good practices for me to bring with me to future projects.

That said, I am always a fan of included suggestions for further self-guided learning specific to the tutorial project. That way if you are having trouble coming up with something yourself, you're not just reading the docs without a goal. So I thought I would share a few that I did:

1. Add the final score to the retry screen
2. Add combos, as suggested in the first gif. I doubled the score earned with every consecutive bounce (i.e. +1, +2, +4, etc)
3. Keep track of the highest combo and display that to the player on the retry screen.
4. Add a label that appears (easy) next to the score or (harder) where the squashed mob was, telling the player how many points earned. (Recall that note about separating UI from the scoring system)
4a. Make that label move, fade, and then disappear, like you see in many games.
5. Add a second mob type with different characteristics (change the speed, color, model, points earned, spawn rate)

Hope that helps someone!

**Reply by marcollivris** [reply]
It helps, thank You. 
I would add: 
 6.   Play a different short sound to underline the death event
      6a.   Maybe stop or pause the main background music at death
 7.   Currently, the restart screen is waiting for any keypress, 
       maybe try to reduce that to an actual ENTER key press
 8.   Add an "Enter your name" label + text input for player name
 9.   Add a Start screen and a Start button
 10. Bored to play ? Want to escape ? Add a ESC key listening, 
       allowing to go back to the Start screen
 11. Add a Highscore button, leading to the Hiscores screen
 12. Add a Quit button on Start screen :)

**Comment by gjones84** [comment]
This tutorial is excellent - it's so valuable to be able to go through the complete process start to finish. I've seen tutorials for other engines/IDEs which start with a half-built project, and while that might be good for quickly getting into learning about a specific technique, you're still pretty much on your own when it comes to actually creating a new project from scratch. I feel much more prepared to start building my own games in Godot from the ground up after completing this tutorial.

**Comment by molorosh** [comment]
This tutorial was really good fun to do, and I feel like I learned a lot! 
A couple of extra things I added were:
a) I added some logic (in the player script) to prevent jumping if the player position.y was greater than 5 which effectively gave me a nice "double-jump" without it going to crazy levels, and
b) I added more logic so that if the player position.y was less than -5 then I called the die() method (which happens if the player goes off the floor and begins to fall down and down (and down (and down)).
c) I also turned the music down by dragging the Volume dB on the music player to about -10 and finally,
d) I noticed that when I died the retry screen was accepting the SPACE key as well and the ENTER key (and SPACE is one of the main controls).  To fix this I went to Project Setting / Input Map and toggled on the defaults.  Then I added another key action for ui_accept and then deleted the space mapping and then my newly added one too - and then the game will only start when you hit the ENTER key 👍

**Comment by FrankCasanova** [comment]
I completed this tutorial using Rust as a scripting language thanks to the https://godot-rust.github.io/ community
good tutorial, cover well the most key-features to start any game.

**Comment by knightmb** [comment]
The tutorials have been great. Coming from my own experience with decades of coding in other game engines along with the same amount of time in game design, it's refreshing to "start over" and learn from the ground up instead of assuming that all your previous knowledge will just fall into place in a new game engine. I also read the comments for every tutorial section steps because it provides a lot of "outside" the tutorial insight from others.

**Comment by yvesevrard-hue** [comment]
Thanks for the tutoriels and mostly the Game Engine. You Pro.

**Comment by Blenhahaha** [comment]
on godot 4.5 I cannot get the death screen to work. Everything else worked perfectly except upon death the game hangs and focus on game window is lost. 
I've gone back and checked my work and cannot find anything i miss-did or miss typed. To double check I even replaced my code with the code given at the end exercise. 

I get one warning on start and 2 errors on death:

warning:
W 0:00:00:416   GDScript::reload: Integer division. Decimal part will be discarded.
  <GDScript Error>INTEGER_DIVISION
  <GDScript Source>mob.gd:31 @ GDScript::reload()

- relating to line 31 in mob.gd

error1:
E 0:00:04:870   main.gd:19 @ _on_mob_timer_timeout(): Node not found: "Player" (relative to "/root/Main").
  <C++ Error>   Method/function failed. Returning: nullptr
  <C++ Source>  scene/main/node.cpp:1908 @ get_node()
  <Stack Trace> main.gd:19 @ _on_mob_timer_timeout()

- relating to line 19 in main.gd

error2:
E 0:00:04:870   _on_mob_timer_timeout: Invalid access to property or key 'position' on a base object of type 'null instance'.
  <GDScript Source>main.gd:19 @ _on_mob_timer_timeout()
  <Stack Trace> main.gd:19 @ _on_mob_timer_timeout()

- relating to line 19 in main.gd

**Reply by KeyMonkeyMonk** [reply]
Did add the code to stop mob timer?
```
func _on_player_hit():
	$MobTimer.stop()
```
The Player is deleted after death, and the mob_time_timeout function at line 19, that your error is referencing, is attempting to access a Player.position. But the Player is null, so it has no position attribute to access, thus the nullptr (null pointer) error.
